File: lingo/interpreters/go/lingolib.go
package lingolib

import (
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

func ParseFile(path string) (map[string]interface{}, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("cannot read %s: %w", path, err)
	}
	var raw map[string]interface{}
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("cannot parse %s: %w", path, err)
	}
	lingo, ok := raw["lingo"].(map[string]interface{})
	if !ok || lingo["spec"] == nil {
		return nil, fmt.Errorf("missing lingo.spec in %s", path)
	}
	return raw, nil
}

func evalExpr(expr interface{}) (string, error) {
	m, ok := expr.(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("unsupported expression type")
	}
	if s, ok := m["str"]; ok {
		return fmt.Sprintf("%v", s), nil
	}
	return "", fmt.Errorf("unsupported expression: %v", m)
}

func ExecuteExe(doc map[string]interface{}) (string, error) {
	main, ok := doc["main"]
	if !ok {
		return "", fmt.Errorf("exe spec missing main")
	}
	return evalExpr(main)
}

func ExecuteFile(path string) (string, error) {
	doc, err := ParseFile(path)
	if err != nil {
		return "", err
	}
	lingo := doc["lingo"].(map[string]interface{})
	spec := fmt.Sprintf("%v", lingo["spec"])
	switch spec {
	case "exe":
		return ExecuteExe(doc)
	default:
		return "", fmt.Errorf("unsupported spec: %s", spec)
	}
}

File: lingo/interpreters/hs/src/Lingolib.hs
{-# LANGUAGE OverloadedStrings #-}
module Lingolib
  ( LingoError (..)
  , parseFile
  , executeExe
  , executeFile
  ) where

import           Data.Text (Text, unpack)
import           Data.Yaml (FromJSON (..), decodeFileEither,
                            prettyPrintParseException, withObject, (.:))


-- lingo.spec envelope

data LingoHeader = LingoHeader
  { lingoSpec    :: Text
  , lingoVersion :: Text
  } deriving (Show)

instance FromJSON LingoHeader where
  parseJSON = withObject "LingoHeader" $ \v ->
    LingoHeader <$> v .: "spec" <*> v .: "version"


-- main expression (currently only {str: ...} supported)

newtype MainExpr = MainExprStr Text
  deriving (Show)

instance FromJSON MainExpr where
  parseJSON = withObject "MainExpr" $ \v ->
    MainExprStr <$> v .: "str"


-- exe script document

data ExeScript = ExeScript
  { exeLingo :: LingoHeader
  , exeMain  :: MainExpr
  } deriving (Show)

instance FromJSON ExeScript where
  parseJSON = withObject "ExeScript" $ \v ->
    ExeScript <$> v .: "lingo" <*> v .: "main"


-- public API

newtype LingoError = LingoError String

instance Show LingoError where
  show (LingoError s) = s

parseFile :: FilePath -> IO (Either LingoError ExeScript)
parseFile path = do
  result <- decodeFileEither path
  return $ case result of
    Left  err -> Left  (LingoError (prettyPrintParseException err))
    Right doc -> Right doc

executeExe :: ExeScript -> String
executeExe script =
  let MainExprStr s = exeMain script
  in unpack s

executeFile :: FilePath -> IO (Either LingoError String)
executeFile path = do
  result <- parseFile path
  return $ case result of
    Left  err    -> Left err
    Right script -> Right (executeExe script)
